### PR TITLE
Add `existsM` and `forallM` to Foldable

### PR DESCRIPTION
**Pull request by refried**
I was missing these utility functions.  I've been out of the loop for a while so let me know if I didn't contrib right, and I'll be happy to update.

### CODE REVIEW

**Comment by kailuowang** [comment]
I think if you use Fordable.Source and `tailRecM` directly rather than `foldM`, you should be able to shortcircuit
e.g. 
```scala
  def existsM[G[_], A](fa: F[A])(p: A => G[Boolean])(implicit G: Monad[G]): G[Boolean] = {
    val src = Foldable.Source.fromFoldable(fa)(self)
    G.tailRecM(src) { srcc => srcc.uncons match {
      case Some((a, src)) => G.map(p(a))(bb => if (bb) Right(true) else Left(src.value))
      case None => G.pure(Right(false))
    }}
  }
```

File: core/src/main/scala/cats/Foldable.scala [hunk 1]
```diff
@@ -327,6 +327,56 @@ import simulacrum.typeclass
...
+    * For example:
+    *
+    * {{{
+    * scala> import cats.implicits._
+    * scala> val F = Foldable[List]
+    * scala> F.existsM(List(1,2,3,4))(n => Option(n <= 4))
+    * res0: Option[Boolean] = Some(true)
+    *
+    * scala> F.existsM(List(1,2,3,4))(n => if (n <= 2) Option(true) else Option(false))
+    * res1: Option[Boolean] = Some(true)
```
Comment (kailuowang) [inline core/src/main/scala/cats/Foldable.scala]: might be better to have a negative test case here, e.g.
```
   * scala> F.existsM(List(1,2,3,4))(n => if (n > 4) Option(true) else Option(false))
   * res1: Option[Boolean] = Some(false)
```

**Comment by refried** [comment]
@kailuowang Yes, but do we want it to short-circuit and skip effects?  I could imagine an argument either way.  For my use case, short-circuiting would probably be fine, but I didn’t want to presume that for everyone.  IIRC scalaz's intentionally doesn’t shortcircuit, but I don’t know the rationale.

**Comment by kailuowang** [comment]
I think there is definitely a use case for short-circuiting implementation. If we have to pick from one of the two, I would argue we should pick the short-circuit one, because the other one isn't that cumbersome to write on use-site if needed. How about we ask for more opinions?

**Comment by refried** [comment]
100% agreement.  I meant to go back and say that short-circuiting would probably be more efficient in my use case; since the effects typically involve extra throwaway computation.

**Comment by refried** [comment]
@kailuowang I was mistaken when I said scalaz doesn't short-circuit; it does:
```
scala> import scalaz._, Scalaz._
import scalaz._
import Scalaz._

scala> List(Some(true), None).anyM(identity)
res0: Option[Boolean] = Some(true)

scala> List(Some(true), None).allM(identity)
res1: Option[Boolean] = None
```
I'll redo the PR when I get a chance.

**Comment by kailuowang** [comment]
err. not sure if you need to close this PR. If you replace with the implementation I gave it shortcircuit and pass the tests.

**Comment by refried** [comment]
Ok, I wasn't sure if I should create a new one, or rebase or what.

File: core/src/main/scala/cats/Foldable.scala [hunk 2]
```diff
@@ -327,6 +327,79 @@ import simulacrum.typeclass
...
+    *
+    * scala> F.forallM(List(1,2,3,4))(n => if (n <= 2) Option(false) else None)
+    * res3: Option[Boolean] = Some(false)
+    *
+    * scala> F.forallM(List(1,2,3,4))(n => if (n <= 2) None else Option(false))
+    * res4: Option[Boolean] = None
+    * }}}
+    */
+  def forallM[G[_], A](fa: F[A])(p: A => G[Boolean])(implicit G: Monad[G]): G[Boolean] = {
+    val src = Foldable.Source.fromFoldable(fa)(self)
```
Comment (johnynek) [inline core/src/main/scala/cats/Foldable.scala]: This isn't being used right.

Comment (refried) [inline core/src/main/scala/cats/Foldable.scala]: I meant to delete line 393, but do you want to elaborate?

**Comment by refried** [comment]
I will add an example to the `tut` docs.  Sorry I didn't think of it sooner.

**Pull request merged**